### PR TITLE
Administration: Add box reordering toggle to Screen Options.

### PR DESCRIPTION
**Pull request by j111q**
Trac ticket: https://core.trac.wordpress.org/ticket/50699

This continues #12180, originally authored by @poligilad-auto. Poli’s original feature commit is preserved in this branch; I’m carrying the existing work and review feedback forward while Poli is away.

## Summary

- Adds a Screen Options setting for enabling or disabling box reordering.
- Stores the setting as one per-user preference shared across screens on the current site and keeps reordering enabled by default.
- Disables drag-and-drop sorting and hides the keyboard move controls and drop zones when reordering is disabled.
- Places the “Allow boxes to be rearranged” checkbox under “Additional settings” on screens that support meta boxes.

## Screenshot

Screenshot 2026-07-22 at 1 05 54 PM

## Review follow-up

- Uses the absence of the disabled body class to represent the enabled state.
- Leaves hidden move controls to CSS instead of changing their accessibility attributes in JavaScript.
- Preserves the existing Screen Options guidance and adds a separate sentence explaining the new setting.
- Uses the user option API consistently when saving and reading the per-user, per-site preference.
- Avoids duplicate sortable updates when the checkbox changes.
- Preserves the legacy hidden editor-expand fieldset when a post screen has no meta boxes.
- Replaces the unused rendering helper test with coverage through the real Screen Options rendering path.

## Compatibility

- Reordering remains enabled by default.
- Existing meta box markup and keyboard controls remain unchanged while reordering is enabled.
- The preference applies across screens on the current site for each user, matching the direction confirmed on the Trac ticket.

## AI assistance

Codex was used to inspect the existing feedback, implement revisions, add tests, and draft this description. I manually verified the user-facing behavior, and the automated checks listed below were run. I do not independently have the technical expertise to validate every code change, so this draft is seeking human developer review before it is considered ready.

## Testing

- `npm run test:php -- --filter Tests_Admin_IncludesScreen tests/phpunit/tests/admin/includesScreen.php`
- `npm run test:php -- --group ajax --filter Tests_Ajax_wpAjaxMetaBoxReordering`
- `npx grunt jshint:core --file=src/js/_enqueues/admin/postbox.js`
- `npx grunt jshint:core --file=src/js/_enqueues/admin/common.js`
- PHPCS on all eight touched PHP and PHPUnit files.
- `npm run build:dev`
- `git diff --check`

Manually verified on the Dashboard that disabling the setting disables drag-and-drop, hides both keyboard move controls, and that re-enabling it restores both interaction methods.

### CODE REVIEW

File: src/wp-admin/includes/ajax-actions.php [hunk 1]
```diff
@@ -1856,6 +1856,20 @@ function wp_ajax_hidden_columns() {
...
+/**
+ * Handles the meta box reordering setting via Ajax.
+ *
+ * @since 7.1.0
+ */
+function wp_ajax_meta_box_reordering() {
+	check_ajax_referer( 'screen-options-nonce', 'screenoptionnonce' );
+
+	$enabled = isset( $_POST['enabled'] ) && '1' === (string) $_POST['enabled'];
+	update_user_option( get_current_user_id(), 'meta_box_reordering', $enabled ? 'enabled' : 'disabled' );
```
[the remark this answers is not in the document]
Comment (j111q) [reply src/wp-admin/includes/ajax-actions.php:1868]: This action is registered only as an authenticated `wp_ajax_` action. Unauthenticated requests use the `wp_ajax_nopriv_` path, for which no handler is registered, and the request also requires the logged-in user’s Screen Options nonce. A user ID of `0` is therefore not reachable through this handler, so I’m keeping this as-is.

File: src/js/_enqueues/admin/postbox.js [hunk 2]
```diff
@@ -437,6 +450,51 @@
...
+				window.wpResponsive.maybeDisableSortables();
+				return;
+			}
+
+			if ( $sortables.hasClass( 'ui-sortable' ) ) {
+				$sortables
+					.sortable( this.metaBoxReorderingEnabled ? 'enable' : 'disable' )
+					.find( '.ui-sortable-handle' )
+						.toggleClass( 'is-non-sortable', ! this.metaBoxReorderingEnabled );
+			}
```
[the remark this answers is not in the document]
Comment (j111q) [reply src/js/_enqueues/admin/postbox.js:495]: All `.meta-box-sortables` elements are initialized together by the `.sortable()` call in `postboxes.init()` immediately before `setMetaBoxReordering()` runs. They therefore share the same initialized state in this code path. Dynamically added elements would not have been initialized by this code either. The current guard also follows the earlier maintainer suggestion, so I’m keeping it as-is.

File: src/wp-admin/includes/ajax-actions.php [hunk 3]
```diff
@@ -1856,6 +1856,20 @@ function wp_ajax_hidden_columns() {
...
+
```
[the remark this answers is not in the document]
Comment (j111q) [reply src/wp-admin/includes/ajax-actions.php:1869]: The intended scope is one preference for a specific user, shared across screens on the current site—not across every site in a multisite network. The default site-specific behavior of `update_user_option()` is therefore intentional. I’ve updated the PR description to make that scope explicit.

File: tests/phpunit/tests/admin/includesScreen.php [hunk 4]
```diff
@@ -260,6 +260,71 @@ public function test_taxonomy_with_special_suffix_as_hookname() {
...
+
+		$this->assertTrue( wp_is_meta_box_reordering_enabled() );
+	}
+
+	public function test_meta_box_reordering_enabled_respects_user_option() {
+		$user_id = self::factory()->user->create( array( 'role' => 'administrator' ) );
+		wp_set_current_user( $user_id );
+		update_user_option( $user_id, 'meta_box_reordering', 'disabled' );
+
+		$this->assertFalse( wp_is_meta_box_reordering_enabled() );
```
[the remark this answers is not in the document]
Comment (j111q) [reply tests/phpunit/tests/admin/includesScreen.php:274]: The intended preference is per user and per site, shared across that site’s screens—not network-wide. This test therefore uses the default site-specific user-option behavior intentionally. I’ve clarified that scope in the PR description.

File: src/js/_enqueues/admin/postbox.js [hunk 5]
```diff
@@ -476,6 +534,30 @@
...
+		save_meta_box_reordering_state : function( enabled ) {
+			$.post(
+				ajaxurl,
+				{
+					action: 'meta-box-reordering',
+					enabled: enabled ? 1 : 0,
+					screenoptionnonce: $( '#screenoptionnonce' ).val()
+				},
+				function() {
+					wp.a11y.speak( __( 'Screen Options updated.' ) );
```
Comment (joedolson) [inline src/js/_enqueues/admin/postbox.js:556]: If this only updates the metabox reordering state, then it should have a more specific message: e.g. "Meta box reordering disabled" or "Meta box reordering enabled".

And all of the reordering text should be updated for consistency; the label was changed to "Allow boxes to be rearranged", and all related text should be consistent with that.

File: src/wp-admin/css/common.css [hunk 6]
```diff
@@ -1982,6 +1982,18 @@ p.auto-update-status {
 	line-height: 2.35;
 }
 
+.metabox-prefs > p {
+	margin: 0 0 8px;
+}
+
+.additional-settings-prefs {
+	margin-top: 14px;
```
Comment (joedolson) [inline src/wp-admin/css/common.css:1990]: This number seems arbitrary; is it based on anything?

File: src/wp-admin/css/common.css [hunk 7]
```diff
@@ -1982,6 +1982,18 @@ p.auto-update-status {
...
+}
+
+.additional-settings-prefs legend {
+	padding-bottom: 2px;
```
Comment (joedolson) [inline src/wp-admin/css/common.css:1994]: This is overridden by styles with higher specificity above, line 1970. Probably can be omitted.

File: src/wp-admin/includes/class-wp-screen.php [hunk 8]
```diff
@@ -1006,11 +1006,31 @@ public function show_screen_options() {
...
+				$additional_settings .= __( 'Enable full-height editor and distraction-free functionality.' ) . '</label>';
+			} else {
+				$expand                 = '<fieldset class="editor-expand hidden"><legend>' . __( 'Additional settings' ) . '</legend><label for="editor-expand-toggle">';
+				$expand                .= '<input type="checkbox" id="editor-expand-toggle"' . checked( get_user_setting( 'editor_expand', 'on' ), 'on', false ) . ' />';
+				$expand                .= __( 'Enable full-height editor and distraction-free functionality.' ) . '</label></fieldset>';
+				$this->_screen_settings = $expand;
+			}
+		}
+
+		if ( $additional_settings ) {
```
Comment (joedolson) [inline src/wp-admin/includes/class-wp-screen.php:1029]: This is unnecessarily complex;it shouldn't be necessary to modify the constructed HTML based on the existence of the meta box reordering option - an additional settings fieldset and legend is needed either way, and it seems like both the $expand setting and the reorder setting should be inserted inside that fieldset.

File: src/wp-admin/includes/class-wp-screen.php [hunk 9]
```diff
@@ -1023,7 +1043,7 @@ public function show_screen_options() {
 		 */
 		$this->_screen_settings = apply_filters( 'screen_settings', $this->_screen_settings, $this );
 
-		if ( $this->_screen_settings || $this->_options ) {
+		if ( $this->_screen_settings || $this->_options || $show_meta_box_reordering_option ) {
```
Comment (joedolson) [inline src/wp-admin/includes/class-wp-screen.php:1046]: What is the case where `$show_meta_box_reordering_option` would not be set, but `$this->_screen_settings` would be? I'm struggling to see why this extra condition is needed. 

If $show_meta_box_reordering_option` is true, then `$additional_settings` is constructed, and if that's constructed, then `$this-_screen_settings` is set.

File: src/wp-admin/includes/class-wp-screen.php [hunk 10]
```diff
@@ -1148,6 +1171,35 @@ public function render_meta_boxes_preferences() {
...
 
+	/**
+	 * Gets the option to enable or disable meta box reordering.
+	 *
+	 * @since 7.1.0
+	 *
+	 * @return string Meta box reordering option markup.
+	 */
+	private function get_meta_box_reordering_option() {
+		return '<label for="meta-box-reordering">'
```
Comment (joedolson) [inline src/wp-admin/includes/class-wp-screen.php:1182]: Update the HTML so it's not an implicit label e.g. so the `label` doesn't wrap the `input`.

**Review by joedolson — CHANGES_REQUESTED** [review]
A handful of minor questions/suggestions.
